File: Interview_Prep_App/native_java/retrofit/src/main/java/com/dreams/interviewprepapp/repositories/Tc2rGithubRepository.java
package com.dreams.interviewprepapp.repositories;

import com.dreams.interviewprepapp.network.RetrofitBuilders;
import com.dreams.interviewprepapp.repositories.models.response.AnswersResponse;
import com.dreams.interviewprepapp.repositories.models.response.QuestionsResponse;
import com.dreams.interviewprepapp.repositories.tc2rgithubrepository.source.remote.Tc2rGithubService;

import retrofit2.Call;

public class Tc2rGithubRepository {
    private static final RetrofitBuilders retrofitBuilder = new RetrofitBuilders();

    private final Tc2rGithubService tc2rGithubService = retrofitBuilder.tc2rGithubService;

    public Call<AnswersResponse> getAnswersList = getSelectedQuizAnswers("ANDROID"); //TODO change this as per user selected quiz language

    public Call<QuestionsResponse> getQuestionsList = getSelectedQuizQuestions("ANDROID"); //TODO change this as per user selected quiz language

    // request for answers json based on selected quiz language
    private Call<AnswersResponse> getSelectedQuizAnswers(String selectedLanguage) {
         Call<AnswersResponse> answersResponseCall;
         switch (selectedLanguage.toUpperCase()) {
            case "ANDROID" : answersResponseCall = tc2rGithubService.getAndroidAnswersList(); break;
            case "C"       : answersResponseCall = tc2rGithubService.getCAnswers()         ;  break;
            case "C#"      : answersResponseCall = tc2rGithubService.getCSharpAnswers()    ;  break;
            case "CPP"     : answersResponseCall = tc2rGithubService.getCPPAnswers()       ;  break;
            case "DART"    : answersResponseCall = tc2rGithubService.getDartAnswers()      ;  break;
            case "DP"      : answersResponseCall = tc2rGithubService.getDPAnswers()        ;  break;
            case "DSA"     : answersResponseCall = tc2rGithubService.getDSAAnswers()       ;  break;
            case "GOLANG"  : answersResponseCall = tc2rGithubService.getGolangAnswers()    ;  break;
            case "HTML5"   : answersResponseCall = tc2rGithubService.getHTML5Answers()     ;  break;
            case "JS"      : answersResponseCall = tc2rGithubService.getJSAnswers()        ;  break;
            case "JAVA"    : answersResponseCall = tc2rGithubService.getJavaAnswers()      ;  break;
            case "PHP"     : answersResponseCall = tc2rGithubService.getPHPAnswers()       ;  break;
            case "PYTHON"  : answersResponseCall = tc2rGithubService.getPythonAnswers()    ;  break;
            case "RUBY"    : answersResponseCall = tc2rGithubService.getRubyAnswers()      ;  break;
            case "IOS"     : answersResponseCall = tc2rGithubService.getIOSAnswers()       ;  break;
            default        : answersResponseCall = tc2rGithubService.getAndroidAnswersList(); break;
        };
        return answersResponseCall;
    }

    // request for questions json based on selected quiz language
    private Call<QuestionsResponse> getSelectedQuizQuestions(String selectedLanguage) {
        Call<QuestionsResponse> questionsResponseCall;
        switch (selectedLanguage.toUpperCase()) {
            case "ANDROID" : questionsResponseCall = tc2rGithubService.getAndroidQuestionsList() ; break;
            case "C"       : questionsResponseCall = tc2rGithubService.getCQuestionsJson()        ; break;
            case "C#"      : questionsResponseCall = tc2rGithubService.getCSharpQuestionsJson()   ; break;
            case "CPP"     : questionsResponseCall = tc2rGithubService.getCPPQuestionsJson()      ; break;
            case "DART"    : questionsResponseCall = tc2rGithubService.getDartQuestionsJson()     ; break;
            case "DP"      : questionsResponseCall = tc2rGithubService.getDPQuestionsJson()       ; break;
            case "DSA"     : questionsResponseCall = tc2rGithubService.getDSAQuestionsJson()      ; break;
            case "GOLANG"  : questionsResponseCall = tc2rGithubService.getGolangQuestionsJson()   ; break;
            case "HTML5"   : questionsResponseCall = tc2rGithubService.getHTML5QuestionsJson()    ; break;
            case "JS"      : questionsResponseCall = tc2rGithubService.getJSQuestionsJson()       ; break;
            case "JAVA"    : questionsResponseCall = tc2rGithubService.getJavaQuestionsJson()     ; break;
            case "PHP"     : questionsResponseCall = tc2rGithubService.getPHPQuestionsJson()      ; break;
            case "PYTHON"  : questionsResponseCall = tc2rGithubService.getPythonQuestionsJson()   ; break;
            case "RUBY"    : questionsResponseCall = tc2rGithubService.getRubyQuestionsJson()     ; break;
            case "IOS"     : questionsResponseCall = tc2rGithubService.getIOSQuestionsJson()      ; break;
            default        : questionsResponseCall = tc2rGithubService.getAndroidQuestionsList()  ; break;
        }
        return questionsResponseCall;
    }

}

File: Interview_Prep_App/native_kotlin/retrofit/src/main/java/com/interviewprep/kotlinretrofit/repository/tc2rgithubrepository/Tc2rGithubRepository.kt
package com.interviewprep.kotlinretrofit.repository.tc2rgithubrepository

import android.util.Log
import com.interviewprep.kotlinretrofit.network.RetrofitBuilders
import com.interviewprep.kotlinretrofit.repository.models.Answer
import com.interviewprep.kotlinretrofit.repository.models.Question
import com.interviewprep.kotlinretrofit.repository.models.response.AnswersResponse
import com.interviewprep.kotlinretrofit.repository.models.response.QuestionsResponse
import com.interviewprep.kotlinretrofit.repository.tc2rgithubrepository.source.remote.Tc2rGithubService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import retrofit2.Response

class Tc2rGithubRepository {
    private val TAG = "ACTEST"

    companion object {
        private val retrofitBuilder: RetrofitBuilders = RetrofitBuilders()
    }

    // Creates API object using Retrofit to implement API Service
    private val tc2rGithubService: Tc2rGithubService by lazy {
        retrofitBuilder.tc2rGithubService
    }

    suspend fun getTheAnswers(): ArrayList<Answer> {
        val answersList: ArrayList<Answer> = arrayListOf()
        withContext(Dispatchers.IO) {
            val response = requestSelectedQuizAnswers("ANDROID")//TODO pass in user selected language
            val array = response.body()

            val responseCode = response.code()
            Log.i(TAG, "$responseCode")

            for (item in array?.answers!!) {
                val tempAnswer = item.answer
                val tempDetail = item.details

                val temp = Answer(tempAnswer, tempDetail)

                Log.i(TAG, "onResponse TempAnswer: $tempAnswer")
                Log.i(TAG, "onResponse TempDetail: $tempDetail")
                answersList.add(temp)
            }
        }
        return answersList
    }

    suspend fun getQuestions(): ArrayList<Question> {
        val questionsList: ArrayList<Question> = arrayListOf()
        withContext(Dispatchers.IO) {
            val response = requestSelectedQuizQuestions("ANDROID")//TODO pass in user selected language
            val array = response.body()
            Log.i(TAG, "onResponse Array: $array")

            for (item in array?.questions!!) {
                val tempId = item.id
                val tempQuestion = item.question
                val tempDetails = item.details
                val tempType = item.questionType
                val tempShortAns = item.shortAns
                val tempTF = item.trueOrFalse

                // Skips boolean type questions. Implement feature later
                if (tempType == "bool") continue

                val temp = Question(
                    item.id,
                    item.questionType,
                    item.question,
                    tempDetails,
                    tempShortAns,
                    tempTF
                )

                Log.i(TAG, "TempQuestion: $tempQuestion $tempId")
                questionsList.add(temp)
            }
        }
        return questionsList
    }

    // request for answers json based on selected quiz language
    suspend fun requestSelectedQuizAnswers(selectedLanguage: String): Response<AnswersResponse> {
        return when (selectedLanguage.uppercase()) {
            "ANDROID" -> tc2rGithubService.requestAndroidAnswers()
            "C"       -> tc2rGithubService.requestCAnswers()
            "C#"      -> tc2rGithubService.requestCSharpAnswers()
            "CPP"     -> tc2rGithubService.requestCPPAnswers()
            "DART"    -> tc2rGithubService.requestDartAnswers()
            "DP"      -> tc2rGithubService.requestDPAnswers()
            "DSA"     -> tc2rGithubService.requestDSAAnswers()
            "GOLANG"  -> tc2rGithubService.requestGolangAnswers()
            "HTML5"   -> tc2rGithubService.requestHTML5Answers()
            "JS"      -> tc2rGithubService.requestJSAnswers()
            "JAVA"    -> tc2rGithubService.requestJavaAnswers()
            "PHP"     -> tc2rGithubService.requestPHPAnswers()
            "PYTHON"  -> tc2rGithubService.requestPythonAnswers()
            "RUBY"    -> tc2rGithubService.requestRubyAnswers()
            "IOS"     -> tc2rGithubService.requestIOSAnswers()
            else      -> tc2rGithubService.requestAndroidAnswers()
        }
    }

    // request for questions json based on selected quiz language
    suspend fun requestSelectedQuizQuestions(selectedLanguage: String): Response<QuestionsResponse> {
        return when (selectedLanguage.uppercase()) {
            "ANDROID" -> tc2rGithubService.getAndroidQuestionsJson()
            "C"       -> tc2rGithubService.getCQuestionsJson()
            "C#"      -> tc2rGithubService.getCSharpQuestionsJson()
            "CPP"     -> tc2rGithubService.getCPPQuestionsJson()
            "DART"    -> tc2rGithubService.getDartQuestionsJson()
            "DP"      -> tc2rGithubService.getDPQuestionsJson()
            "DSA"     -> tc2rGithubService.getDSAQuestionsJson()
            "GOLANG"  -> tc2rGithubService.getGolangQuestionsJson()
            "HTML5"   -> tc2rGithubService.getHTML5QuestionsJson()
            "JS"      -> tc2rGithubService.getJSQuestionsJson()
            "JAVA"    -> tc2rGithubService.getJavaQuestionsJson()
            "PHP"     -> tc2rGithubService.getPHPQuestionsJson()
            "PYTHON"  -> tc2rGithubService.getPythonQuestionsJson()
            "RUBY"    -> tc2rGithubService.getRubyQuestionsJson()
            "IOS"     -> tc2rGithubService.getIOSQuestionsJson()
            else      -> tc2rGithubService.getAndroidQuestionsJson()
        }
    }
}
